File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.cs
using System;
using System.Collections.Generic;

public class Solution {
    public string[] solution(string[] players, string[] callings) {
       int length = players.Length;

Dictionary<string, int> playerDict = new Dictionary<string, int>();

for(int i = 0; i < length; i++)
{
    playerDict.Add(players[i], i);
}

foreach(string calling in callings) 
{
    string loserName = players[playerDict[calling] - 1];

    players[playerDict[calling] - 1] = players[playerDict[calling]];
    players[playerDict[calling]] = loserName;

    int temp = playerDict[calling];
    playerDict[calling] = playerDict[loserName];
    playerDict[loserName] = temp;
}

for(int i = 0; i < length; i++)
{
    int changeIndex = playerDict[players[i]];

    string temp = players[i];
    players[i] = players[changeIndex];
    players[changeIndex] = temp;
}

return players;
    }
}

File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.py
def solution(players, callings):
    answer = []

    players_dict = {name: index for index, name in enumerate(players)}

    for calling in callings:
        winner_index = players_dict[calling]
        loser_index = players_dict[calling] - 1

        players_dict[players[loser_index]] += 1
        players_dict[calling] -= 1

        players[loser_index], players[winner_index] = players[winner_index], players[loser_index]

    answer = list(dict(sorted(players_dict.items(), key=lambda x: x[1])).keys())
    
    return answer